nixos/modules/system/boot/osk-sdl.nix

```diff
@@ -0,0 +1,92 @@
+{ config, lib, pkgs, utils, ... }:
+
+with lib;
+
+let
+  cfg = config.boot.initrd.osk-sdl;
+in
+{
+  options.boot.initrd.osk-sdl = {
+    enable = mkEnableOption (lib.mdDoc "osk-sdl in initrd") // {
+      description = lib.mdDoc ''
+        Whether to enable the osk-sdl on-screen keyboard in initrd to unlock LUKS.
+      '';
+    };
+  };
+
+  config = {
+    meta.maintainers = with maintainers; [ tomfitzhenry ];
+    assertions = [
+      {
+        assertion = cfg.enable -> config.boot.initrd.systemd.enable;
+        message = "boot.initrd.osk-sdl is only supported with boot.initrd.systemd.";
+      }
+    ];
+    boot.initrd.availableKernelModules = [
+      # See also:
+      # https://salsa.debian.org/DebianOnMobile-team/osk-sdl/-/blob/debian/master/debian/initramfs/hooks/osk-sdl#L57
+      # https://wiki.archlinux.org/title/kernel_mode_setting#Early_KMS_start
+
+      "evdev" # for entering pw
+      "virtio-gpu" # for qemu
+    ];
+
+    boot.initrd.systemd = {
+      storePaths = with pkgs; [
+        dejavu_fonts
+        gnugrep
+        libGL
+        mesa.drivers
+        osk-sdl
+        "${systemd}/lib/systemd/systemd-reply-password"
+      ];
+      services = {
+        osk-sdl-ask-password = {
+          description = "Forward Password Requests to OSK-SDL";
+          conflicts = [
+            "shutdown.target"
+            "emergency.service"
+          ];
+          unitConfig.DefaultDependencies = false;
+          after = [
+            "systemd-vconsole-setup.service"
+          ];
+          before = [
+            "shutdown.target"
+          ];
+          script = ''
+            mkdir -p /run/opengl-driver/
+            cp -r ${pkgs.mesa.drivers}/* /run/opengl-driver/
+
+            export SDL_VIDEODRIVER=kmsdrm
+            for file in `ls /run/systemd/ask-password/*`; do
+              socket="$(cat "$file" | ${pkgs.gnugrep}/bin/grep "Socket=" | cut -d= -f2)"
+              ${pkgs.osk-sdl}/bin/osk-sdl -c ${pkgs.osk-sdl}/etc/osk.conf -d x -n x -k -s | ${pkgs.systemd}/lib/systemd/systemd-reply-password 1 "$socket"
+            done
```

Comment:
Looking at https://github.com/systemd/systemd/blob/main/src/tty-ask-password-agent/tty-ask-password-agent.c, this appears to be a very incomplete way to respond to password requests.

Reply:
Also, the way this script is written, if a new password request is queued after the loop has begun, it won't be answered at all by this unit, I think. Unless the below `path` unit will restart this if the directory continues to be non-empty? I don't think it does though.